File: self pased DSA/Hashing/L12_Subarray_with_given_sum.cpp
#include<bits/stdc++.h>
#include<unordered_set>
using namespace std;
// Naive method, O(n^2) time
bool isSumZero(int arr[],int n,int x)
{
    
    for (int i = 0; i < n; i++)
    {
        int sum1=0;
        if(arr[i]==0) {return true;}
        for (int j = i; j < n; j++)
        {
            sum1=sum1+arr[j];
            if(sum1==x)
                return true;
        }
        
    }
    return false;
    
}
// Effective approch
// using prefix sum and hashing  
bool isZeroSubarraySum(int arr[],int n,int sum)
{
    unordered_set<int> h;
    int pre_sum=0;
    for (int i = 0; i < n; i++)
    {
        pre_sum +=arr[i];

        if(pre_sum==sum)    // corner case
            return true;
        if(h.find(pre_sum-sum)!=h.end())
            return true;
        else
            h.insert(pre_sum);
    }
    return false;
}
int main(){
    int arr[]={1,4,13,-3,-10,5};
    int sum=14;
    int n=sizeof(arr)/sizeof(arr[0]);
    cout<<isSumZero(arr,n,sum)<<endl;
    cout<<isZeroSubarraySum(arr,n,sum)<<endl;
    return 0;
}

File: code_java/self pased DSA/Hashing/L12_Subarray_with_given_sum.java
/*
Problem: Check if there exists a subarray with a given sum in an array.
Approach:
1. Naive: Check all subarrays. O(n^2)
2. Efficient: Use prefix sum and HashSet to check for the sum in O(n).
*/
import java.util.*;

public class L12_Subarray_with_given_sum {
    // Naive method
    static boolean isSumZero(int[] arr, int x) {
        for (int i = 0; i < arr.length; i++) {
            int sum1 = 0;
            if (arr[i] == 0) return true;
            for (int j = i; j < arr.length; j++) {
                sum1 += arr[j];
                if (sum1 == x) return true;
            }
        }
        return false;
    }
    // Efficient method
    static boolean isZeroSubarraySum(int[] arr, int sum) {
        Set<Integer> h = new HashSet<>();
        int pre_sum = 0;
        for (int value : arr) {
            pre_sum += value;
            if (pre_sum == sum) return true;
            if (h.contains(pre_sum - sum)) return true;
            h.add(pre_sum);
        }
        return false;
    }
    public static void main(String[] args) {
        int[] arr = {1, 4, 13, -3, -10, 5};
        int sum = 14;
        System.out.println(isSumZero(arr, sum));
        System.out.println(isZeroSubarraySum(arr, sum));
    }
} 